### PR TITLE
Implement DHCP for machine network

### PR DESCRIPTION
**Pull request by mcbenjemaa**
THis PR adds:

- Support DHCP

### CODE REVIEW

File: api/v1alpha1/helpers.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2023 IONOS Cloud.
```
[the remark this answers is not in the document]
Comment (wikkyk) [reply api/v1alpha1/helpers.go:2]: wondering if the license action will complain :)

File: api/v1alpha1/proxmoxcluster_types.go [hunk 2]
```diff
@@ -50,25 +48,49 @@ type ProxmoxClusterSpec struct {
...
 	// DNSServers contains information about nameservers used by machines network-config.
 	// +kubebuilder:validation:MinItems=1
 	DNSServers []string `json:"dnsServers"`
 }
 
+// IPConfig contains information about available IP config.
+type IPConfig struct {
+	// Addresses is a list of IP addresses that can be assigned. This set of
+	// addresses can be non-contiguous.
+	// +kubebuilder:validation:MinItems=1
```
Comment (avorima) [inline api/v1alpha1/proxmoxcluster_types.go:77]: Shouldn't DHCP and addresses be mutually exclusive?

File: api/v1alpha1/proxmoxcluster_types.go [hunk 3]
```diff
@@ -50,25 +48,49 @@ type ProxmoxClusterSpec struct {
...
+	// +optional
+	Prefix int `json:"prefix,omitempty"`
+
+	// Gateway
+	// +optional
+	Gateway string `json:"gateway,omitempty"`
+
+	// DHCP indicates if DHCP should be used to assign IP addresses.
+	// +optional
+	DHCP bool `json:"dhcp,omitempty"`
```
Comment (avorima) [inline api/v1alpha1/proxmoxcluster_types.go:91]: This needs to be a pointer if `false` is supposed to be a valid value

**Review by avorima — CHANGES_REQUESTED** [review]
First batch of comments. Didn't look at vmservice yet

File: api/v1alpha1/helpers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (lubedacht) [reply api/v1alpha1/helpers.go:2]: Maybe use https://github.com/denis-tingaikin/go-header in [golangci-lint](https://golangci-lint.run/usage/linters/#goheader).

You can define a year range there, and it would also make the action obsolete.

File: api/v1alpha1/proxmoxcluster_types.go (hunk 2 above)
Comment (65278) [reply api/v1alpha1/proxmoxcluster_types.go:77]: DHCP and static IP addresses on teh same interface aren't mutually exclusive, and they also aren't in netplan.

Comment (avorima) [reply api/v1alpha1/proxmoxcluster_types.go:77]: Ok, let me rephrase. Should addresses really be required when using DHCP is an option?

File: api/v1alpha1/helpers.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	if !c.HasMachine(loc.Machine.Name, isControlPlane) {
+		c.addNodeLocation(loc, isControlPlane)
+	}
+}
+
+// RemoveNodeLocation removes a node location from the status.
+func (c *ProxmoxCluster) RemoveNodeLocation(machineName string, isControlPlane bool) {
+	nodeLocations := c.Status.NodeLocations
+
+	if nodeLocations == nil {
```
[the remark this answers is not in the document]
Comment (mcbenjemaa) [reply api/v1alpha1/helpers.go:65]: but, this is used in `DeleteVM`

[the remark this answers is not in the document]
Comment (mcbenjemaa) [reply api/v1alpha1/helpers.go:65]: right, but that can be removed

File: api/v1alpha1/proxmoxcluster_types.go [hunk 5]
```diff
@@ -50,25 +48,49 @@ type ProxmoxClusterSpec struct {
...
 
+// IPConfig contains information about available IP config.
+type IPConfig struct {
+	// Addresses is a list of IP addresses that can be assigned. This set of
+	// addresses can be non-contiguous.
+	// +kubebuilder:validation:MinItems=1
+	Addresses []string `json:"addresses,omitempty"`
+
+	// Prefix is the network prefix to use.
+	// +kubebuilder:validation:Maximum=128
```
[the remark this answers is not in the document]
Comment (mcbenjemaa) [reply api/v1alpha1/proxmoxcluster_types.go:81]: I don't understand 😄 

I didn't change anything here!

the webhook already check this

File: api/v1alpha1/proxmoxcluster_types.go (hunk 2 above)
Comment (mcbenjemaa) [reply api/v1alpha1/proxmoxcluster_types.go:77]: No, it's mutually exclusive.

File: api/v1alpha1/helpers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (wikkyk) [reply api/v1alpha1/helpers.go:2]: I've deleted the action for now. We can evaluate go-header and if we like it, I'll add the updated licence action back in.

But yes this should say 2024.

File: api/v1alpha1/proxmoxmachine_types.go [hunk 6]
```diff
@@ -223,12 +223,22 @@ type NetworkDevice struct {
...
+	// +optional
+	DHCP6 bool `json:"dhcp6,omitempty"`
 }
 
 // AdditionalNetworkDevice the definition of a Proxmox network device.
-// +kubebuilder:validation:XValidation:rule="self.ipv4PoolRef != null || self.ipv6PoolRef != null",message="at least one pool reference must be set, either ipv4PoolRef or ipv6PoolRef"
+// +kubebuilder:validation:XValidation:rule="(self.ipv4PoolRef != null || self.ipv6PoolRef != null || self.dhcp4 || self.dhcp6)",message="at least dhcp and/or one pool reference must be set, either ipv4PoolRef or ipv6PoolRef"
 type AdditionalNetworkDevice struct {
-	NetworkDevice `json:",inline"`
+	*NetworkDevice `json:",inline"`
```
[the remark this answers is not in the document]
Comment (mcbenjemaa) [reply api/v1alpha1/proxmoxmachine_types.go:241]: right

File: internal/controller/proxmoxcluster_controller.go [hunk 7]
```diff
@@ -243,3 +244,16 @@ func (r *ProxmoxClusterReconciler) SetupWithManager(ctx context.Context, mgr ctr
 			builder.WithPredicates(predicates.ClusterUnpaused(ctrl.LoggerFrom(ctx)))).
 		Complete(r)
 }
+
+func dhcpEnabled(cluster *infrav1alpha1.ProxmoxCluster) bool {
```
[the remark this answers is not in the document]
Comment (mcbenjemaa) [reply internal/controller/proxmoxcluster_controller.go:248]: yes

File: api/v1alpha1/proxmoxcluster_types.go (hunk 3 above)
Comment (mcbenjemaa) [reply api/v1alpha1/proxmoxcluster_types.go:91]: but, this needs to be by default `false`

File: api/v1alpha1/helpers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mcbenjemaa) [reply api/v1alpha1/helpers.go]: Please, do that in a separate PR.

File: api/v1alpha1/proxmoxcluster_types.go (hunk 3 above)
Comment (avorima) [reply api/v1alpha1/proxmoxcluster_types.go:94]: You don't have to set a default. You can just treat nil like false

**Comment by wikkyk** [comment]
Rebased and updated copyright headers.

File: api/v1alpha1/proxmoxcluster_types.go [hunk 8]
```diff
@@ -50,25 +48,52 @@ type ProxmoxClusterSpec struct {
...
+	// +optional
+	Prefix int `json:"prefix,omitempty"`
+
+	// Gateway
+	// +optional
+	Gateway string `json:"gateway,omitempty"`
+
+	// DHCP indicates that if DHCP should be used to assign IP addresses.
```
Comment (wikkyk) [inline api/v1alpha1/proxmoxcluster_types.go:90]: DHCP indicates that ~if~ DHCP should be used to assign IP addresses.

File: internal/service/vmservice/bootstrap.go [hunk 9]
```diff
@@ -40,9 +40,9 @@ func reconcileBootstrapData(ctx context.Context, machineScope *scope.MachineScop
 		return false, nil
 	}
 
-	if !machineHasIPAddress(machineScope.ProxmoxMachine) {
-		// skip machine doesn't have an IpAddress yet.
-		conditions.MarkFalse(machineScope.ProxmoxMachine, infrav1alpha1.VMProvisionedCondition, infrav1alpha1.WaitingForStaticIPAllocationReason, clusterv1.ConditionSeverityWarning, "no ip address")
+	if machineScope.VirtualMachine == nil {
+		// skip machine is not yet provisioned.
```
Comment (wikkyk) [inline internal/service/vmservice/bootstrap.go:44]: skip machine if it is not provisioned yet

**Review by wikkyk — CHANGES_REQUESTED** [review]

**Comment by wikkyk** [comment]
This PR should also update the docs as appropriate.

File: api/v1alpha1/proxmoxcluster_types.go (hunk 3 above)
Comment (mcbenjemaa) [reply api/v1alpha1/proxmoxcluster_types.go:94]: so, i will have to change this fn and will be more complicated

```
// DHCPEnabled returns whether DHCP is enabled.
func (c ClusterNetworkConfig) DHCPEnabled() bool {
	switch {
	case (c.IPv6Config != nil && c.IPv6Config.DHCP) && (c.IPv4Config != nil && c.IPv4Config.DHCP):
		return true
	case (c.IPv6Config != nil && c.IPv6Config.DHCP) && c.IPv4Config == nil:
		return true
	case (c.IPv4Config != nil && c.IPv4Config.DHCP) && c.IPv6Config == nil:
		return true
	default:
		return false
	}
}

```

File: internal/service/vmservice/ip.go [hunk 10]
```diff
@@ -74,13 +74,13 @@ func formatIPAddressName(name, device string) string {
 	return fmt.Sprintf("%s-%s", name, device)
 }
 
-func machineHasIPAddress(machine *infrav1alpha1.ProxmoxMachine) bool {
-	return machine.Status.IPAddresses[infrav1alpha1.DefaultNetworkDevice] != (infrav1alpha1.IPAddress{})
+func machineHasIPAddress(machine *infrav1.ProxmoxMachine) bool {
+	return machine.Status.IPAddresses[infrav1.DefaultNetworkDevice] != (infrav1.IPAddress{})
 }
 
 func handleIPAddressForDevice(ctx context.Context, machineScope *scope.MachineScope, device, format string, ipamRef *corev1.TypedLocalObjectReference) (string, error) {
```
Comment (wikkyk) [inline internal/service/vmservice/ip.go:81]: Can this function be refactored so that it always returns an error when it also returns an empty string? To avoid all the `if err != nil || ip == "" {`.

File: internal/service/vmservice/vm.go [hunk 11]
```diff
@@ -241,7 +242,9 @@ func reconcileMachineAddresses(scope *scope.MachineScope) error {
 	return nil
 }
 
-func getMachineAddresses(scope *scope.MachineScope) ([]clusterv1.MachineAddress, error) {
+// getMachineAddresses returns the ip addresses for the machine
+//nolint
```
Comment (wikkyk) [inline internal/service/vmservice/vm.go:246]: why not?

**Review by wikkyk — CHANGES_REQUESTED** [review]
First round of reviews.

File: api/v1alpha1/proxmoxcluster_types.go (hunk 3 above)
Comment (avorima) [reply api/v1alpha1/proxmoxcluster_types.go:94]: Use `ptr.Deref` from `k8s.io/utils`

```go
// DHCPEnabled returns whether DHCP is enabled.
func (c ClusterNetworkConfig) DHCPEnabled() bool {
	switch {
	case (c.IPv6Config != nil && ptr.Deref(c.IPv6Config.DHCP, false)) && (c.IPv4Config != nil && ptr.Deref(c.IPv4Config.DHCP, false)):
		return true
	case (c.IPv6Config != nil && ptr.Deref(c.IPv6Config.DHCP, false)) && c.IPv4Config == nil:
		return true
	case (c.IPv4Config != nil && ptr.Deref(c.IPv4Config.DHCP, false)) && c.IPv6Config == nil:
		return true
	default:
		return false
	}
}
```

The bool pointer makes the DHCP setting clearer from the user's perspective because the explicit `false` stays `false`.

Comment (avorima) [reply api/v1alpha1/proxmoxcluster_types.go:94]: Actually, to reduce dependencies you can just copy the ptr.Deref function into the api package. Keeps imports clean

File: api/v1alpha1/proxmoxcluster_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (avorima) [reply api/v1alpha1/proxmoxcluster_types.go:82]: You're right, it's a bit out of scope for this PR. The webhook doesn't validate that IPv4 only gets prefix 32 though

File: internal/service/vmservice/ip.go (hunk 10 above)
Comment (mcbenjemaa) [reply internal/service/vmservice/ip.go:81]: It's very confusing to return an error when creating a new IP claim:

https://github.com/ionos-cloud/cluster-api-provider-proxmox/blob/0cd195bb54ec8ddd57c80061676d77e82fc1d6d0/internal/service/vmservice/ip.go#L94-L98

 also when trying to add a tag:

https://github.com/ionos-cloud/cluster-api-provider-proxmox/blob/0cd195bb54ec8ddd57c80061676d77e82fc1d6d0/internal/service/vmservice/ip.go#L112-L117

Another choice is to add a third return as bool, but it's actually the same.

File: api/v1alpha1/proxmoxcluster_types.go (hunk 3 above)
Comment (mcbenjemaa) [reply api/v1alpha1/proxmoxcluster_types.go]: yes, you can check

File: .github/workflows/lint.yml
[the remark this answers is not in the document]
Comment (mcbenjemaa) [reply .github/workflows/lint.yml:1]: https://github.com/ionos-cloud/cluster-api-provider-proxmox/pull/101

**Comment by mcbenjemaa** [comment]
I'm ready

File: api/v1alpha1/proxmoxcluster_types.go (hunk 2 above)
Comment (avorima) [reply api/v1alpha1/proxmoxcluster_types.go:78]: Ok. It can't be mutually exclusive if the addresses are always expected to have one item

File: internal/service/vmservice/vm.go [hunk 12]
```diff
@@ -257,21 +259,70 @@ func getMachineAddresses(scope *scope.MachineScope) ([]clusterv1.MachineAddress,
...
+	var ipv4, ipv6 string
+	// get ip addresses from vm if dhcp is enabled
+	if hasDHCPEnabled(scope.InfraCluster.ProxmoxCluster.Spec.IPv4Config,
+		networkSpec.Default,
+		infrav1alpha1.DefaultNetworkDevice,
+		infrav1alpha1.IPV4Format) ||
+		hasDHCPEnabled(scope.InfraCluster.ProxmoxCluster.Spec.IPv6Config,
+			networkSpec.Default,
+			infrav1alpha1.DefaultNetworkDevice,
+			infrav1alpha1.IPV6Format) {
```
Comment (avorima) [inline internal/service/vmservice/vm.go:277]: It would actually be less code if you move the default interface handling out of `hasDHCPEnabled` and use a few intermediate variables, e.g.

```go
if dev := networkSpec.Default; scope.InfraCluster.ProxmoxCluster.Spec.DHCPEnabled() || (dev != nil && dev.DHCP4 || dev.DHCP6) {
```

But more importantly, why is the default device DHCP setting even considered when DHCP is enabled on the cluster?

File: internal/service/vmservice/vm.go [hunk 13]
```diff
@@ -257,21 +259,70 @@ func getMachineAddresses(scope *scope.MachineScope) ([]clusterv1.MachineAddress,
...
+		}
+
+		for _, net := range nets {
+			// default device
+			mac := extractMACAddress(scope.VirtualMachine.VirtualMachineConfig.Net0)
+			if strings.EqualFold(mac, net.HardwareAddress) {
+				for _, ip := range net.IPAddresses {
+					if ip.IPAddressType == "ipv4" && ip.IPAddress != scope.InfraCluster.ProxmoxCluster.Spec.ControlPlaneEndpoint.Host {
+						ipv4 = ip.IPAddress
+					} else if ip.IPAddressType == "ipv6" && !strings.Contains(ip.IPAddress, "fe80:") {
```
Comment (avorima) [inline internal/service/vmservice/vm.go:290]: Please use [`netip.Addr.IsLinkLocalUnicast`](https://pkg.go.dev/net/netip#Addr.IsLinkLocalUnicast). Substring match for IPv6 is always less accurate than comparing the bits.

File: api/v1alpha1/proxmoxmachine_types.go [hunk 14]
```diff
@@ -224,10 +224,20 @@ type NetworkDevice struct {
 	// +kubebuilder:validation:Minimum=1
 	// +kubebuilder:validation:Maximum=65520
 	MTU *uint16 `json:"mtu,omitempty"`
+
+	// DHCP4 indicates that if DHCP should be used to assign IPv4 addresses.
+	// DHCP4 enforce device to use DHCP despite the config set in cluster.spec.ipv4Config.
+	// +optional
+	DHCP4 bool `json:"dhcp4,omitempty"`
```
Comment (avorima) [inline api/v1alpha1/proxmoxmachine_types.go:231]: Why is in the base network device? All permuations of DHCP and IPAM on the default device should be configured in the cluster network config.

This also has the same pointer issue as the cluster dhcp setting.

File: api/v1alpha1/proxmoxcluster_types.go [hunk 15]
```diff
@@ -50,25 +48,52 @@ type ProxmoxClusterSpec struct {
...
+	Prefix int `json:"prefix,omitempty"`
+
+	// Gateway
+	// +optional
+	Gateway string `json:"gateway,omitempty"`
+
+	// DHCP indicates that DHCP should be used to assign IP addresses.
+	// mutually exclusive with Addresses.
+	// +optional
+	// +kubebuilder:default=false
```
Comment (avorima) [inline api/v1alpha1/proxmoxcluster_types.go:93]: This default isn't needed. The behavior stays the same whether it's set to `false` or `null`. My pointer request was about making the user-facing API explicit and minimizing changes.

File: pkg/proxmox/goproxmox/api_client.go [hunk 16]
```diff
@@ -245,3 +245,13 @@ func (c *APIClient) StartVM(ctx context.Context, vm *proxmox.VirtualMachine) (*p
 func (c *APIClient) TagVM(ctx context.Context, vm *proxmox.VirtualMachine, tag string) (*proxmox.Task, error) {
 	return vm.AddTag(ctx, tag)
 }
+
+// GetVMNetwork returns a VM network interfaces based on nodeName and vmID.
+func (c *APIClient) GetVMNetwork(ctx context.Context, vm *proxmox.VirtualMachine) (iFaces []*proxmox.AgentNetworkIface, err error) {
```
Comment (avorima) [inline pkg/proxmox/goproxmox/api_client.go:250]: nit: named return values are not used and don't provide any additional context

**Review by avorima — COMMENTED** [review]
Sorry, for the long wait.

File: api/v1alpha1/proxmoxcluster_types.go (hunk 2 above)
Comment (mcbenjemaa) [reply api/v1alpha1/proxmoxcluster_types.go:78]: That's already considered via webhook.

File: api/v1alpha1/helpers.go [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+	case (c.IPv6Config != nil && ptr.Deref(c.IPv6Config.DHCP, false)) && (c.IPv4Config != nil && ptr.Deref(c.IPv4Config.DHCP, false)):
+		return true
+	case (c.IPv6Config != nil && ptr.Deref(c.IPv6Config.DHCP, false)) && c.IPv4Config == nil:
+		return true
+	case (c.IPv4Config != nil && ptr.Deref(c.IPv4Config.DHCP, false)) && c.IPv6Config == nil:
+		return true
+	default:
+		return false
+	}
+}
```
[the remark this answers is not in the document]
Comment (mcbenjemaa) [reply api/v1alpha1/helpers.go:168]: The problem here is we need to check if the cluster has dhcp so we can skip the IPAM pool.

Then, in the machine, we also need to check if the machine has DHCP.

[the remark this answers is not in the document]
Comment (avorima) [reply api/v1alpha1/helpers.go:168]: This doesn't contradict my proposition

File: api/v1alpha1/proxmoxcluster_types.go (hunk 2 above)
Comment (65278) [reply api/v1alpha1/proxmoxcluster_types.go:78]: Coming back to this, last week I debugged a setup which required static ip + dhcp. It's a valid usecase I guess.

Comment (avorima) [reply api/v1alpha1/proxmoxcluster_types.go:78]: I just noticed that my initial question came from a wrong assumption that the marker would force me to always provide addresses. The behavior is already how I would've wanted it to be. Sorry for the hold up
